pushsnapscript/snap_store.py

```diff
@@ -2,48 +2,220 @@
+    'edge': VersionType.BETA,
+    'beta': VersionType.BETA,
+    'candidate': VersionType.RELEASE,
+    'stable': VersionType.RELEASE,
+
+    'esr': VersionType.ESR,
+    'esr/stable': VersionType.ESR,
+    'esr/candidate': VersionType.ESR,
+}
+
+# TODO: Parametrize this once we have other products
+_SNAP_NAME_ON_STORE = 'firefox'
+
+
 def push(context, snap_file_path, channel):
     if not task.is_allowed_to_push_to_snap_store(channel=channel):
-        log.warn('Not allowed to push to Snap store. Skipping push...')
+        log.warning('Not allowed to push to Snap store. Skipping push...')
         # We don't raise an error because we still want green tasks on dev instances
         return
 
     macaroon_location = context.config['macaroons_locations'][channel]
-    with _session(macaroon_location):
-        push_kwargs = {
-            'snap_filename': snap_file_path,
-            'release_channels': [channel],
-        }
-        log.debug('Calling snapcraft push with these kwargs: {}'.format(push_kwargs))
-        snapcraft_store_client.push(**push_kwargs)
+    with _store_session(macaroon_location) as store:
+        try:
+            log.debug('Calling snapcraft.push() with this file: {}'.format(snap_file_path))
+            # We don't call store.upload() because this push() does more
+            snapcraft_store_client.push(snap_filename=snap_file_path)
+        except StoreReviewError as e:
+            if 'A file with this exact same content has already been uploaded' in e.additional:
+                log.warning(e)
+            else:
+                raise
+
+        _release_if_needed(store, channel, snap_file_path)
 
 
 @contextlib.contextmanager
-def _session(macaroon_location):
-    store = storeapi.StoreClient()
+def _store_session(macaroon_location):
+    store = StoreClient()
 
     with open(macaroon_location) as macaroon:
         log.debug('Logging onto Snap store with macaroon file "{}"...'.format(macaroon.name))
-        snapcraft_store_client.login(store=store, config_fd=macaroon)
+        # XXX Bad API. email and password are mandatory in the function signature, but they are
+        # read from the macaroon when config_fd is provided
+        store.login(email='', password='', config_fd=macaroon)
 
     log.info('Logged on Snap store')
     try:
-        yield
+        yield store
     finally:
         log.debug('Logging off Snap store...')
         store.logout()
         log.info('Logged off Snap store')
+
+
+def _release_if_needed(store, channel, snap_file_path):
+    all_revisions = _list_all_revisions(store)
+    metadata_per_revision = _pluck_metadata(all_revisions)
+    metadata_per_revision = _filter_versions_that_are_not_the_same_type(metadata_per_revision, channel)
+    metadata_per_revision = _populate_sha3_384(store, metadata_per_revision)
+    current_sha3_384 = get_hash(snap_file_path, hash_alg='sha3_384')
+    current_snap_revision, current_snap_version = _find_revision_and_version_of_current_snap(
```

Comment:
I'm not following the logic here. Won't this raise if the current snap *isn't* already uploaded? Or are we running `populate_sha3_384` to actually push some metadata to the store? If so, what happens if the metadata for that version is already there? How about if the version is there, and the sha doesn't match?

I think this patch could use more docstrings.

Reply:
You're right it will raise if not uploaded. At this stage, the snap must be uploaded. This function is about releasing an uploaded (but unreleased) snap. There is no metadata to push except "point this channel to this revision". 

The last question is a great one. I'll follow up on it. That's a scenario I hadn't thought of.

I'll also add more docstrings for context. Thank you for bringing it all up!

Reply:
Docsstrings added.